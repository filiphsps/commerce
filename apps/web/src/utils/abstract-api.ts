/* eslint-disable unused-imports/no-unused-vars */
import type { Shop } from '@/api/shop';
import type { Locale } from '@/utils/locale';
import type { ApolloClient, FetchPolicy, TypedDocumentNode } from '@apollo/client';
import { DocumentTransform } from '@apollo/client';
import { visit } from 'graphql';
import { unstable_cache } from 'next/cache';

export type Optional<T extends { [key: string]: unknown }> = { [K in keyof T]?: Nullable<T[K]> };
export type Nullable<T> = T | null;
export type Identifiable = { handle: string };

export type OmitTypeName<T> = Omit<T, '__typename'>;

export type ApiOptions = { api: AbstractApi };

export type AbstractApi<Q = any> = {
    locale: () => Locale;
    shop: () => Shop;
    query: <T>(
        query: Q,
        variables?: Record<string, string | number | boolean | object | Array<string | number | object> | null>,
        options?: {
            fetchPolicy?: FetchPolicy;
            tags?: string[];
            revalidate?: number;
        }
    ) => Promise<{ data: T | null; errors: readonly any[] | undefined }>;
};
export type AbstractApiBuilder<K, Q> = ({
    api,
    locale,
    shop,
    fetchPolicy
}: {
    api: K;
    locale: Locale;
    shop: Shop;
    fetchPolicy?: FetchPolicy;
}) => AbstractApi<Q>;

export type AbstractShopifyApolloApiBuilder<Q> = AbstractApiBuilder<ApolloClient<any>, Q>;

/**
 * Creates an AbstractApiBuilder for Shopify Apollo APIs.
 *
 * @todo TODO: Improve the type safety of all `AbstractApi` implementations.
 *
 * @param {object} options - The api options.
 * @param {ApolloClient<any>} options.api - The Apollo client to use.
 * @param {Locale} options.locale - The locale to use.
 * @param {FetchPolicy | undefined} options.fetchPolicy - The fetch policy to use.
 * @param {string[] | undefined} options.tags - The nextjs fetch tags to use.
 * @returns {AbstractApiBuilder} The AbstractApiBuilder.
 */
export const ApiBuilder: AbstractShopifyApolloApiBuilder<TypedDocumentNode<any, any>> = ({
    api,
    locale,
    shop,
    fetchPolicy = 'cache-first'
}) => ({
    locale: () => locale,
    shop: () => shop,
    query: async (query, variables = {}, { tags = [], revalidate = undefined, fetchPolicy = undefined } = {}) => {
        const processedTags = [
            'shopify',
            ...tags.map((tag) => `shopify.${tag}`),
            `shopify.${shop.id}`,
            `shopify.${shop.id}.${locale.code}`,
            ...tags.map((tag) => `shopify.${shop.id}.${locale.code}.${tag}`)
        ];

        const { data, errors, error } = await api.query({
            query,
            fetchPolicy,
            context: {
                fetchOptions: {
                    cache: fetchPolicy || revalidate ? undefined : 'force-cache',
                    next: {
                        revalidate,
                        tags: processedTags
                    }
                }
            },
            variables: {
                language: locale.language,
                country: locale.country,
                ...variables
            }
        });

        return { data: data || null, errors, error };
    }
});

export const cache: typeof unstable_cache = (func, keyparts, options) => {
    /*if (typeof (React as any).cache === 'undefined') {
        console.warn('React cache is unavailable in this environment.');
        return func;
    }*/

    if (typeof unstable_cache === 'undefined') {
        console.warn('next/cache is unavailable in this environment.');
        return func;
    }

    return unstable_cache(func, keyparts, options);
};

/**
 * @todo TODO: This should be replaced with a generalized shopify parser.
 *       Preferably one that we can use to output in whatever format we want.
 */
export const cleanShopifyHtml = (html: string | unknown): Nullable<string> => {
    if (typeof html !== 'string' || !html) return null;
    let out = html as string;

    // Remove all non-breaking spaces and replace them with normal spaces.
    // TODO: This is a hacky solution. We should write a proper shopify parser.
    out = out.replaceAll(/ /g, ' ').replaceAll('\u00A0', ' ');

    // Replace some of the more common unicode characters with their HTML.
    out = out
        .replaceAll('”', '&rdquo;')
        .replaceAll('“', '&ldquo;')
        .replaceAll('‘', '&lsquo;')
        .replaceAll('’', '&rsquo;')
        .replaceAll('…', '&hellip;');

    // Trim the preceding and trailing whitespace.
    out = out.trim();

    return out;
};

/**
 * A GraphQL document transform that adds the `@inContext` directive to all queries.
 *
 * THIS IS (or at least seems to be) A REALLY HACKY SOLUTION. USE WITH CAUTION. :^)
 *
 * @todo TODO: This should be placed in a separate file or even a separate package.
 * @todo TODO: Need to add tests for this.
 * @todo TODO: This currently adds duplicate variables if they already exist.
 */
export const shopifyContextTransform = new DocumentTransform((document) => {
    // Add Shopify's `@inContext` directive to all queries.
    const transformedDocument = visit(document, {
        OperationDefinition(node) {
            return {
                ...node,
                variableDefinitions: [
                    ...(node.variableDefinitions || []),
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'country'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'CountryCode'
                            }
                        }
                    },
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'language'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'LanguageCode'
                            }
                        }
                    }
                ],
                directives: [
                    ...(node.directives || []),
                    {
                        kind: 'Directive',
                        name: {
                            kind: 'Name',
                            value: 'inContext'
                        },
                        arguments: [
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'country'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'country'
                                    }
                                }
                            },
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'language'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'language'
                                    }
                                }
                            }
                        ]
                    }
                ]
            };
        }
    });
    return transformedDocument;
});
