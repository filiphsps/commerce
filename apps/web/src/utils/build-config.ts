export const BuildConfig = {
    environment: process.env.NODE_ENV || 'production',
    domain: process.env.DOMAIN || 'www.sweetsideofsweden.com',
    GTM: process.env.GTM,

    shopify: {
        shop_id: (process.env.SHOPIFY_SHOP_ID && Number.parseInt(process.env.SHOPIFY_SHOP_ID)) || 76188483889,
        storefront_id: process.env.SHOPIFY_STOREFRONT_ID || '2130225',
        domain: (process.env.SHOPIFY_DOMAIN || 'sweet-side-of-sweden.myshopify.com').replace('https://', ''),
        token: process.env.SHOPIFY_TOKEN || '319eb651b3464ea882a016ca2085ebc1',
        private_token: process.env.SHOPIFY_PRIVATE_TOKEN || undefined,
        api: '2023-10',
        checkout_domain: process.env.SHOPIFY_CHECKOUT_DOMAIN || 'checkout.sweetsideofsweden.com'
    },

    prismic: {
        toolbar: false,
        repo: process.env.PRISMIC_REPO || 'https://candy-by-sweden.cdn.prismic.io/api/v2',
        name: (process.env.PRISMIC_REPO || 'candy-by-sweden')
            .replace('https://', '')
            .replace('.cdn.prismic.io/api/v2', '')
    },

    i18n: {
        locales: process.env.STORE_LOCALES?.split?.(',') || ['en-US'],
        default: process.env.STORE_DEFAULT_LOCALE || 'en-US',
        currencies: process.env.STORE_CURRENCIES?.split(',') || ['USD']
    }
};
