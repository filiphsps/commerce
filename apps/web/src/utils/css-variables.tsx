import { ShopApi } from '@/api/shop';
import { colord } from 'colord';
import { unstable_cache as cache } from 'next/cache';

// TODO: Generalize this
export const getBrandingColors = async (domain: string) => {
    return cache(
        async (domain: string) => {
            const shop = await ShopApi(domain);
            const { colors } = shop.configuration!.design!.branding!;

            // TODO: Deal with variants.
            const primary = colors!.find(({ type }) => type === 'primary')!;
            const secondary = colors!.find(({ type }) => type === 'secondary')!;

            return {
                primary,
                secondary
            };
        },
        [domain, 'branding']
    )(domain);
};

const CssVariablesProvider = async ({ domain }: { domain: string }) => {
    const branding = await getBrandingColors(domain);

    // TODO: Background and foreground colors.
    return (
        <style suppressHydrationWarning={true}>{`
        :root {
            --color-background: #fefefe;
            --color-foreground: #101418;

            --color-accent-primary: ${branding.primary.accent};
            --color-accent-primary-text: ${branding.primary.foreground};
            --color-accent-primary-light: ${colord(branding.primary.accent).lighten(0.175).saturate(0.15).toHex()};
            --color-accent-primary-dark: ${colord(branding.primary.accent).darken(0.05).toHex()};

            --color-accent-secondary: ${branding.secondary.accent};
            --color-accent-secondary-text: ${branding.secondary.foreground};
            --color-accent-secondary-light: ${colord(branding.secondary.accent).lighten(0.15).toHex()};
            --color-accent-secondary-dark: ${colord(branding.secondary.accent).darken(0.15).toHex()};

            /* TODO: Remove these legacy variables. */
            --accent-primary: var(--color-accent-primary);
            --accent-primary-light: var(--color-accent-primary-light);
            --accent-primary-dark: var(--color-accent-primary-dark);
            --accent-secondary: var(--color-accent-secondary);
            --accent-secondary-light: var(--color-accent-secondary-light);
            --accent-secondary-dark: var(--color-accent-secondary-dark);
        }
    `}</style>
    );
};

CssVariablesProvider.displayName = 'Nordcom.CssVariablesProvider';
export { CssVariablesProvider };
