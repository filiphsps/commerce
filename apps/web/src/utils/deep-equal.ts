export function deepEqual<T>(a: T, b: T): boolean {
    try {
        return JSON.stringify(a, null, 0) === JSON.stringify(b, null, 0);
    } catch {
        return false;
    }
}
