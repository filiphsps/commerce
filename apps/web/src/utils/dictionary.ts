import 'server-only';

import type { Shop } from '@/api/shop';
import type { LanguageCode, Locale, LocaleDictionary } from '@/utils/locale';

const stub = { cart: {}, common: {} };

export const dictionaries: Record<Lowercase<LanguageCode>, () => Promise<LocaleDictionary>> = {
    af: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ak: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    am: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ar: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    as: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    az: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    be: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    bg: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    bm: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    bn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    bo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    br: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    bs: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ca: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ce: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ckb: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    cs: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    cu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    cy: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    da: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    de: () => import('@/i18n/de.json').then((module) => module.default) as any,
    dz: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ee: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    el: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    en: () => import('@/i18n/en.json').then((module) => module.default) as any,
    eo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    es: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    et: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    eu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fa: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ff: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fi: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fil: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fr: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    fy: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ga: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    gd: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    gl: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    gu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    gv: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ha: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    he: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    hi: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    hr: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    hu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    hy: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ia: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    id: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ig: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ii: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    is: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    it: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ja: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    jv: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ka: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ki: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    kk: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    kl: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    km: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    kn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ko: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ks: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ku: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    kw: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ky: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    la: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lb: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lg: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ln: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lt: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    lv: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mg: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mi: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mk: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ml: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mr: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ms: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    mt: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    my: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    nb: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    nd: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ne: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    nl: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    nn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    no: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    om: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    or: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    os: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    pa: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    pl: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ps: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    pt_br: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    pt_pt: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    pt: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    qu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    rm: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    rn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ro: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ru: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    rw: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sa: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sc: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sd: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    se: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sg: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sh: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    si: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sk: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sl: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    so: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sq: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sr: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    su: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    sv: () => import('@/i18n/sv.json').then((module) => module.default) as any,
    sw: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ta: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    te: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    tg: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    th: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ti: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    tk: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    to: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    tr: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    tt: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ug: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    uk: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    ur: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    uz: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    vi: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    vo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    wo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    xh: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    yi: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    yo: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    zh_cn: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    zh_tw: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    zh: () => new Promise((resolve) => resolve(stub)) as any /* TODO */,
    zu: () => new Promise((resolve) => resolve(stub)) as any /* TODO */
};

export type DictionaryLanguageCode = keyof typeof dictionaries;

/**
 * Get dictionary for locale.
 * @todo Handle templates.
 *
 * @param {Locale} locale - Locale to get dictionary for.
 * @returns {Promise<LocaleDictionary>} Promise with dictionary.
 */
export const getDictionary = async (props: { shop: Shop; locale: Locale } | Locale): Promise<LocaleDictionary> => {
    let locale: Locale, shop: Shop | undefined;

    if (Object.hasOwn(props, 'shop')) {
        const temp = props as { shop: Shop; locale: Locale };
        locale = temp.locale;
        // eslint-disable-next-line unused-imports/no-unused-vars
        shop = temp.shop;
    } else {
        locale = props as Locale;
    }

    // TODO: Fetch tenant-specific dictionary if it exists and shop is defined.
    return dictionaries[locale?.language?.toLowerCase() as DictionaryLanguageCode]?.() ?? {};
};
