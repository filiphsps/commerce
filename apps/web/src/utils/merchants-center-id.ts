import type { Locale } from '@/utils/locale';
import { parseGid } from '@shopify/hydrogen-react';

/**
 * This function is used to generate a unique ID for a product in the Merchants Center.
 *
 * @param {Locale} locale - The locale of the product.
 * @param {string} productId - The Shopify GID of the product.
 * @param {string} variantId - The Shopify GID of the variant.
 * @returns {string} A unique ID for the product in the Merchants Center.
 */
export const ProductToMerchantsCenterId = ({
    locale,
    product
}: {
    locale: Locale;
    product: {
        productGid: string;
        variantGid: string;
    };
}) => {
    const productId = parseGid(product.productGid).resourceId;
    const variantId = parseGid(product.variantGid).resourceId;

    return `shopify_${locale.country?.toUpperCase()}_${productId}_${variantId}`;
};
