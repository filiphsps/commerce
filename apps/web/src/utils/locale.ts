import type { CountryCode, CurrencyCode, LanguageCode } from '@shopify/hydrogen-react/storefront-api-types';

import { BuildConfig } from '@/utils/build-config';
import type { StoreModel } from '@/models/StoreModel';

// TODO: This should be tenant configurable.
const defaultLocale = BuildConfig.i18n.default;

export type { CountryCode, CurrencyCode, LanguageCode };

export type Locale = {
    locale: string; // xx-XX
    language: LanguageCode;
    country: CountryCode;
    currency?: CurrencyCode;
};

/***
 * Converts a locale string to a `CountryCode`.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {CountryCode} `CountryCode` string.
 */
export const NextLocaleToCountry = (locale?: string): CountryCode =>
    (locale?.split('-')?.[1] || defaultLocale.split('-')[1]).toUpperCase() as CountryCode;

/***
 * Converts a locale string to a `LanguageCode`.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {LanguageCode} `LanguageCode` string.
 */
export const NextLocaleToLanguage = (locale?: string): LanguageCode =>
    (
        (locale && locale.length === 2 && locale) ||
        locale?.split('-')?.[0] ||
        defaultLocale.split('-')[0]
    ).toUpperCase() as LanguageCode; // TODO: replace `toUpperCase` with `toLowerCase`

/***
 * Converts a locale string to a `CurrencyCode`.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {CurrencyCode} `CurrencyCode` string.
 */
export const NextLocaleToCurrency = ({ country, store }: { country: CountryCode; store: StoreModel }): CurrencyCode =>
    (store?.payment?.countries?.find(({ isoCode }) => isoCode === country)?.currency.isoCode ||
        BuildConfig.i18n.currencies[0]) as CurrencyCode;

/***
 * Converts a locale string to a Locale.
 *
 * > NOTE: If the locale is invalid, the default locale will be used.
 *         the default locale is defined in `Config.i18n.default` and
 *         is not tenant configurable at the moment.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {Locale} `Locale` object.
 */
export const NextLocaleToLocale = (locale?: string): Locale | null => {
    if (
        !locale ||
        locale === 'x-default' ||
        locale.length < 2 ||
        locale.length > 5 ||
        (locale.length > 2 && !locale.includes('-'))
    ) {
        // FIXME: Handle invalid locales in a better way.
        return null;
    }

    if (locale.length === 2) {
        // FIXME: Get default country for a given language.
        throw new Error('Not implemented');
    }

    const country = NextLocaleToCountry(locale);
    const language = NextLocaleToLanguage(locale);
    return {
        locale: `${language.toLowerCase()}-${country}`,
        language,
        country
        // TODO: Add currency.
    };
};

/***
 * Returns the default locale.
 *
 * @returns {Locale} `Locale` object.
 */
export const DefaultLocale = (): Locale => {
    return NextLocaleToLocale(defaultLocale)!;
};

// TODO: Make this a proper type that somehow reads from the dictionary files?
export type LocaleDictionary = {} & any;

/***
 * Returns a translation function for a given scope and dictionary.
 *
 * > NOTE: This is a very simple implementation of a translation function.
 *
 * @param {string} scope - The scope of the translation.
 * @param {LocaleDictionary} dictionary - The dictionary to use for the translation.
 * @returns {({ t: (key: string) => string })} The translation function.
 */
export const useTranslation = (scope: string, dictionary: LocaleDictionary) => {
    return {
        t: (key: string): string => dictionary?.[scope]?.[key] || key
    };
};
