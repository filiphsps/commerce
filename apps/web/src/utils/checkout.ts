import type { Shop } from '@/api/shop';
import type { Locale } from '@/utils/locale';
import { ProductToMerchantsCenterId } from '@/utils/merchants-center-id';
import { sendGTMEvent } from '@next/third-parties/google';
import type { CartWithActions } from '@shopify/hydrogen-react';

// Const hacky workaround for ga4 cross-domain
// Ugly hack taken from StackOverflow
export const getCrossDomainLinkerParameter = () => {
    // create form element, give it an action, make it hidden and prevent the submit event
    const formNode = document.createElement('form') as any;
    formNode.action = 'https://opensource.sweetsideofsweden.com';
    formNode.style.opacity = '0';
    formNode.addEventListener('submit', (event: any) => {
        event.preventDefault();
    });

    // create a button node, make it type=submit and append it to the form
    const buttonNode = document.createElement('button') as any;
    buttonNode.type = 'submit';
    formNode.append(buttonNode);

    // append the form (and button) to the DOM
    document.body.append(formNode);

    // trigger a click on the button node to submit the form
    buttonNode.click();

    // check for the input[name=_gl] hidden input in the form (if decoration worked)
    const _glNode = formNode.querySelector('input[name="_gl"]') as any;

    if (_glNode) return _glNode.value as string;
    return null; // TODO: Maybe throw an error here.
};

export const Checkout = async ({ shop, locale, cart }: { shop: Shop; locale: Locale; cart: CartWithActions }) => {
    if (!cart.totalQuantity || cart.totalQuantity <= 0 || !cart.lines) throw new Error('Cart is empty!');
    else if (!cart.checkoutUrl) throw new Error('Cart is missing checkoutUrl');

    let url = cart.checkoutUrl;
    if (shop.configuration.commerce.type === 'shopify') {
        // Replace xxx.myshopify.com with `shop.configuration.commerce.domain` in an url
        url = url.replace(/[A-Za-z0-9\-\_]+.\.myshopify\.com/, `${shop.configuration.commerce.domain}`);
    }

    try {
        // Google Tracking
        sendGTMEvent({
            ecommerce: null
        });
        sendGTMEvent({
            event: 'begin_checkout',
            ecommerce: {
                currency: cart.cost?.totalAmount?.currencyCode!,
                value: Number.parseFloat(cart.cost?.totalAmount?.amount!),
                items: cart.lines.map(
                    (line) =>
                        line && {
                            item_id: ProductToMerchantsCenterId({
                                locale: locale,
                                product: {
                                    productGid: line.merchandise!.product!.id,
                                    variantGid: line.merchandise!.id
                                } as any
                            }),
                            item_name: line.merchandise?.product?.title,
                            item_variant: line.merchandise?.title,
                            item_brand: line.merchandise?.product?.vendor,
                            currency: line.merchandise?.price?.currencyCode!,
                            price: Number.parseFloat(line.merchandise?.price?.amount!) || undefined,
                            quantity: line.quantity
                        }
                )
            }
        });

        // Microsoft Ads tracking
        if ((window as any).uetq) {
            (window as any).uetq.push('event', 'begin_checkout', {
                ecomm_prodid: cart.lines.map((line) => line && line.merchandise?.id),
                ecomm_pagetype: 'cart',
                ecomm_totalvalue: Number.parseFloat(cart.cost?.totalAmount?.amount! || '0'),
                revenue_value: Number.parseFloat(cart.cost?.totalAmount?.amount! || '0'),
                currency: cart.cost?.totalAmount?.currencyCode!,
                items: cart.lines.map(
                    (line) =>
                        line && {
                            id: line.merchandise?.id,
                            quantity: line.quantity,
                            price: Number.parseFloat(line.merchandise?.price?.amount! || '0')
                        }
                )
            });
        }
    } catch {}

    const ga4 = getCrossDomainLinkerParameter();
    const finalUrl = `${url}${(ga4 && `${(url.includes('?') && '&') || '?'}_gl=${ga4}`) || ''}`;
    window.location.href = finalUrl;
};
