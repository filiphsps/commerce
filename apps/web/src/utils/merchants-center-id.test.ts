import { NextLocaleToLocale } from '@/utils/locale';
import { ProductToMerchantsCenterId } from '@/utils/merchants-center-id';
import { describe, expect, it } from 'vitest';

describe('Utils', () => {
    describe('MerchantsCenterId', () => {
        it(`should generate a Merchants Center product ID based on the product's Shopify GIDs`, () => {
            const product = {
                productGid: 'gid://shopify/Product/123',
                variantGid: 'gid://shopify/ProductVariant/456'
            } as any;

            const locale = NextLocaleToLocale('en-US')!;

            const merchantsCenterId = ProductToMerchantsCenterId({
                locale,
                product
            });

            expect(merchantsCenterId).toBe('shopify_US_123_456');
            expect(merchantsCenterId).toMatchSnapshot();
        });

        it(`should generate a Merchants Center product ID based on the product's Shopify GIDs`, () => {
            const product = {
                productGid: 'gid://shopify/Product/123',
                variantGid: 'gid://shopify/ProductVariant/456'
            } as any;

            const locale = NextLocaleToLocale('fr-CA')!;

            const merchantsCenterId = ProductToMerchantsCenterId({
                locale,
                product
            });

            expect(merchantsCenterId).toBe('shopify_CA_123_456');
            expect(merchantsCenterId).toMatchSnapshot();
        });
    });
});
