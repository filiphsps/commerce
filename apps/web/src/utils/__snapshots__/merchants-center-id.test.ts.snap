// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Utils > MerchantsCenterId > should generate a Merchants Center product ID based on the product's Shopify GIDs 1`] = `"shopify_US_123_456"`;

exports[`Utils > MerchantsCenterId > should generate a Merchants Center product ID based on the product's Shopify GIDs 2`] = `"shopify_CA_123_456"`;
